### PR TITLE
Add .closest and .closestAncestorOrAncestorChild helpers for raw.Element

### PR DESCRIPTION
**Pull request by er1c**
Updated & now includes JSDOM-based tests

### CODE REVIEW

File: js/src/main/scala/fm/common/rich/RichElement.scala [hunk 1]
```diff
@@ -75,6 +75,17 @@ final class RichElement(val elem: Element) extends AnyVal {
   
   def removeClass(className: String): Unit = elem.classList.remove(className)
 
+  /** Toggle between two classes, e.g. ({{{icon.toggleClass("glyphicon-eye-close", "glyphicon-eye-open")}}})*/
+  def toggleClass(valueA: String, valueB: String): Unit = {
+    if (elem.classList.exists{ _ === valueA }) {
```
Comment (er1c) [inline js/src/main/scala/fm/common/rich/RichElement.scala]: this should just use the `hasClass`

File: js/src/main/scala/fm/common/rich/RichNode.scala [hunk 2]
```diff
@@ -29,16 +26,6 @@ final class RichNode[T <: Node](val self: T) extends AnyVal {
   /** Detatch this node from the DOM */
   def detatch(): T = if (hasParent) self.parentNode.removeChild(self).asInstanceOf[T] else self
   
-  /** Find the closest ancestor of the current element (or the current element itself) that matches a class. */
-  def closest[A <: Node : ClassTag]: A = closestImpl(self)
```
Comment (tpunder) [inline js/src/main/scala/fm/common/rich/RichNode.scala]: What's wrong with this implementation?

File: js/src/main/scala/fm/common/rich/RichElement.scala [hunk 3]
```diff
@@ -45,4 +74,86 @@ final class RichElement(val elem: Element) extends AnyVal {
...
+      throw new NoSuchElementException(s"No Matching Parent for ${classTag[A]}")
+    }
+  }
+
+
+
+  /** Toggle between two attribute values, e.g. ({{{passwordField.toggleAttributeValue("type", "text", "password")}}})*/
+  def toggleAttributeValue(attrName: String, valueA: String, valueB: String): Unit = {
+    if (elem.getAttribute(attrName) === valueA) elem.setAttribute(attrName, valueB)
+    else elem.setAttribute(attrName, valueA)
```
Comment (tpunder) [inline js/src/main/scala/fm/common/rich/RichElement.scala]: How about a single `elem.setAttribute` call?

```scala
val newValue: String = if (elem.getAttribute(attrName) === valueA) valueB else valueA
elem.setAttribute(attrName, newValue)
```

or 

```scala
elem.setAttribute(if (elem.getAttribute(attrName) === valueA) valueB else valueA)
```

File: js/src/main/scala/fm/common/rich/RichElement.scala [hunk 4]
```diff
@@ -45,4 +74,86 @@ final class RichElement(val elem: Element) extends AnyVal {
...
+    elem.childNodes.collect {
+      case e: Element => e
+    }
+  }
+
+  /** Find the closest ancestor of the current element (or the current element itself) that matches a class AND selector. */
+  def closest[A <: Element : ClassTag](selector: String): A = closestImpl(elem, selector)
+
+  @tailrec
+  private def closestImpl[A <: Element : ClassTag](element: Element, selector: String): A = {
```
Comment (tpunder) [inline js/src/main/scala/fm/common/rich/RichElement.scala]: Why does this require a selector?

File: js/src/main/scala/fm/common/rich/RichElement.scala [hunk 5]
```diff
@@ -45,4 +74,86 @@ final class RichElement(val elem: Element) extends AnyVal {
...
+    if (classTag[A].runtimeClass.isInstance(element) && element.matches(selector)) element.asInstanceOf[A]
+    else if (element.hasParent) closestImpl[A](element.getParentElement.get, selector)
+    else throw new NoSuchElementException(s"No Matching Parent for ${classTag[A]}")
+  }
+
+  /** Find the closest ancestor or ancestor's child of the current element that matches a class AND selector. */
```
Comment (tpunder) [inline js/src/main/scala/fm/common/rich/RichElement.scala]: I'm not sure what this means.  Can you add examples to the documentation?

The `closestAncestorOrAncestorChildImpl` code checks `elem` and then `elem`'s child elements and then goes up to the parent?  That doesn't seem to match up with this comment.

File: js/src/main/scala/fm/common/rich/RichElement.scala [hunk 6]
```diff
@@ -30,7 +55,11 @@ final class RichElement(val elem: Element) extends AnyVal {
 
   /** Shortcut for window.getComputedStyle(elem) */
   def computedStyle: CSSStyleDeclaration = window.getComputedStyle(elem)
-  
+
+  def hasParent: Boolean = elem.parentNode.isNotNull
+
+  def hasChild: Boolean = elem.childNodes.isNotNull && elem.childNodes.nonEmpty
```
Comment (tpunder) [inline js/src/main/scala/fm/common/rich/RichElement.scala]: Is `hasChildren` a better name?

Comment (tpunder) [reply js/src/main/scala/fm/common/rich/RichElement.scala]: Or maybe these should be `hasParentNode` and `hasChildNodes` to be clearer

File: js/src/main/scala/fm/common/rich/RichNode.scala (hunk 2 above)
Comment (er1c) [reply js/src/main/scala/fm/common/rich/RichNode.scala]: the lack of the selector, I was having an implicit conflict if this was still around vs having one with a selector also as an implicit . I can see if I can isolate the specific issue

File: js/src/main/scala/fm/common/rich/RichElement.scala (hunk 5 above)
Comment (er1c) [reply js/src/main/scala/fm/common/rich/RichElement.scala]: `/** Find the closest ancestor or ancestors' children (of the current element) that matches a class AND selector. (e.g. closestAncestorOrAncestorChild("input[type='password']") will find the closest password input field that was defined before the current element) */`

Is that better, or still obtuse as a name?  `closest` only checks ancestors, it doesn't check their children.

Comment (tpunder) [reply js/src/main/scala/fm/common/rich/RichElement.scala]:
```
html
  body
    div.one
      div.two <-- 1. calling closestAncestorOrAncestorChild[Div](".three") will return the child div.three or the sibling?
        div.three
      input   <-- 2. calling closestAncestorOrAncestorChild[HTMLInputElement]("input") will return this same element?
      div.two <-- 3. calling closestAncestorOrAncestorChild[Div]("div") will return this element?
      div.three
```

 - 1 - Returns the child?
 - 2 - Returns itself?
 - 3 - Returns itself or its parent?

**Comment by tpunder** [comment]
```
[warn] /Users/tim/fm/fm-common/js/src/main/scala/fm/common/rich/RichElement.scala:160:32: abstract type A is unchecked since it is eliminated by erasure
[warn]           if (next.isInstanceOf[A] && elementMatchesTypeAndSelector(next.asInstanceOf[A], selector)) return next.asInstanceOf[A]
```

**Pull request merged**